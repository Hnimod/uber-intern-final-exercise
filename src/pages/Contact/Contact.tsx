import { Formik, Form, FormikHelpers } from 'formik';
import { IoMailSharp } from 'react-icons/io5';

import BackgroundImage from '../../layouts/BackgroundImage';
import { FormButton } from '../../shared/Buttons';
import InputField from './InputField';
import SelectField from './SelectField';
import TextareaField from './TextareaField';

import styles from './Contact.module.scss';

interface Values {
  name: string;
  type: string;
  message: string;
}

const Contact = () => {
  return (
    <BackgroundImage>
      <Formik
        initialValues={{ name: '', type: 'rider', message: '' }}
        onSubmit={(
          values: Values,
          { setSubmitting }: FormikHelpers<Values>
        ) => {
          alert(JSON.stringify(values, null, 2));
          setSubmitting(false);
        }}
      >
        <Form className={styles.form}>
          <div className={styles.title}>
            <IoMailSharp />
            <h1>Contact us</h1>
          </div>
          <div className={styles.detail}>
            <div className={styles.info}>
              <InputField />
              <SelectField />
            </div>
            <div className={styles.message}>
              <TextareaField />
            </div>
          </div>
          <div className={styles.button}>
            <FormButton />
          </div>
        </Form>
      </Formik>
    </BackgroundImage>
  );
};

export default Contact;
