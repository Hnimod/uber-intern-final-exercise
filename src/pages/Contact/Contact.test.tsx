import renderer from 'react-test-renderer';
import { render, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Contact from './Contact';

jest.mock('../../layouts/BackgroundImage', () => 'BackgroundImage');

test('match snapshot', () => {
  const tree = renderer.create(<Contact />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('input field render correctly', () => {
  const { queryByTestId } = render(<Contact />);
  const input = queryByTestId('contactInput');
  expect(input).toBeTruthy();
});

test('show error text when not enter input field', async () => {
  const { queryByTestId, getByText } = render(<Contact />);
  const input = queryByTestId('contactInput') as HTMLInputElement;
  fireEvent.blur(input);
  let err;
  await waitFor(() => {
    err = getByText('Please enter your name');
  });
  expect(err).toBeTruthy();
});

test('input field recieve correct value from user', async () => {
  const { queryByTestId } = render(<Contact />);
  const input = queryByTestId('contactInput') as HTMLInputElement;
  await userEvent.type(input, 'Minh', { delay: 1 });
  expect(input.value).toBe('Minh');
});

test('data submit correctly', async () => {
  const handleSubmit = jest.fn();
  const { queryByTestId, queryByRole } = render(
    <Contact onSubmitForm={handleSubmit} />
  );
  const input = queryByTestId('contactInput') as HTMLInputElement;
  const select = queryByTestId('contactSelect') as HTMLSelectElement;
  const textarea = queryByTestId('contactTextarea') as HTMLTextAreaElement;
  const button = queryByRole('button') as HTMLButtonElement;
  await userEvent.type(input, 'Minh', { delay: 1 });
  await userEvent.selectOptions(select, ['rider']);
  await userEvent.type(textarea, 'Complain', { delay: 1 });
  userEvent.click(button);

  await waitFor(() => {
    expect(handleSubmit).toHaveBeenCalledTimes(1);
    expect(handleSubmit).toHaveBeenCalledWith({
      name: 'Minh',
      type: 'rider',
      message: 'Complain',
    });
  });
});
