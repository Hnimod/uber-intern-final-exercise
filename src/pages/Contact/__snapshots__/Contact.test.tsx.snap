// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<BackgroundImage>
  <form
    action="#"
    className="form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="title"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M464 80H48a16 16 0 00-16 16v320a16 16 0 0016 16h416a16 16 0 0016-16V96a16 16 0 00-16-16zM265.82 284.63a16 16 0 01-19.64 0L89.55 162.81l19.64-25.26L256 251.73l146.81-114.18 19.64 25.26z"
        />
      </svg>
      <h1>
        Contact us
      </h1>
    </div>
    <div
      className="detail"
    >
      <div
        className="info"
      >
        <section
          className="nameInput"
        >
          <div
            className="input "
          >
            <label
              htmlFor="name"
            >
              Your name:
            </label>
            <input
              data-testid="contactInput"
              id="name"
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Enter your name"
              value=""
            />
          </div>
        </section>
        <section
          className="typeSelect"
        >
          <div
            className="select "
          >
            <label
              htmlFor="type"
            >
              Select:
            </label>
            <select
              data-testid="contactSelect"
              id="type"
              name="type"
              onBlur={[Function]}
              onChange={[Function]}
              value="select"
            >
              <option
                data-testid="option-select"
                value="select"
              >
                Select
              </option>
              <option
                data-testid="option-rider"
                value="rider"
              >
                Rider
              </option>
              <option
                data-testid="option-driver"
                value="driver"
              >
                Driver
              </option>
            </select>
          </div>
        </section>
      </div>
      <div
        className="message"
      >
        <section
          className="messageArea"
        >
          <div
            className="textarea "
          >
            <label
              htmlFor="message"
            >
              Message:
            </label>
            <textarea
              data-testid="contactTextarea"
              id="message"
              name="message"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Enter your message"
              rows={4}
              value=""
            />
          </div>
        </section>
      </div>
    </div>
    <div
      className="button"
    >
      <button
        className="formButton undefined"
        disabled={true}
        type="submit"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 5l7 7m0 0l-7 7m7-7H3"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
      </button>
    </div>
  </form>
</BackgroundImage>
`;
